File: visitor/java/GeometricShapes.java
interface Graphic {

	void accept(Visitor v);
}

interface Visitor {

	void visitShape(Shape shape);

	void visitDot(Dot dot);

	void visitRectangle(Rectangle rectangle);

	void visitCircle(Circle circle);
}

final class ConcreteVisitor implements Visitor {

	@Override
	public void visitShape(Shape shape) {
		System.out.println("Shape visited");
	}

	@Override
	public void visitDot(Dot dot) {
		System.out.println("Dot visited");
	}

	@Override
	public void visitRectangle(Rectangle rectangle) {
		System.out.println("Rectangle visited");
	}

	@Override
	public void visitCircle(Circle circle) {
		System.out.println("Circle visited");
	}
}

class Shape implements Graphic {

	@Override
	public void accept(Visitor v) {
		v.visitShape(this);
	}
}

class Dot extends Shape {

	@Override
	public void accept(Visitor v) {
		v.visitDot(this);
	}
}

class Rectangle extends Dot {

	@Override
	public void accept(Visitor v) {
		v.visitRectangle(this);
	}
}

final class Circle extends Shape {

	@Override
	public void accept(Visitor v) {
		v.visitCircle(this);
	}
}

public final class GeometricShapes {

	public static void main(String[] args) {
		Visitor visitor = new ConcreteVisitor();
		Graphic[] graphics = new Graphic[4];
		graphics[0] = new Shape();
		graphics[1] = new Dot();
		graphics[2] = new Rectangle();
		graphics[3] = new Circle();

		for (Graphic g : graphics) g.accept(visitor);
	}
}

File: visitor/kotlin/geometric_shapes.kt
interface Visitor {

	fun visitShape(shape: Shape): Unit

	fun visitDot(dot: Dot): Unit

	fun visitRectangle(rectangle: Rectangle): Unit

	fun visitCircle(circle: Circle): Unit
}

interface Graphic {

	fun accept(v: Visitor): Unit
}

open class Shape() : Graphic {

	override fun accept(v: Visitor): Unit {
		v.visitShape(this)
	}
}

open class Dot() : Shape() {

	override fun accept(v: Visitor): Unit {
		v.visitDot(this)
	}
}

class Rectangle() : Dot() {

	override fun accept(v: Visitor): Unit {
		v.visitRectangle(this)
	}
}

class Circle() : Graphic {

	override fun accept(v: Visitor): Unit {
		v.visitCircle(this)
	}
}

class ConcreteVisitor() : Visitor {

	override fun visitShape(shape: Shape): Unit {
		println("Shape visited")
	}

	override fun visitDot(dot: Dot): Unit {
		println("Dot visited")
	}

	override fun visitRectangle(rectangle: Rectangle): Unit {
		println("Rectangle visited")
	}

	override fun visitCircle(circle: Circle): Unit {
		println("Circle visited")
	}
}

fun main(): Unit {
	val visitor: Visitor = ConcreteVisitor()
	val graphics: Array<Graphic?> = Array(4) { null }
	graphics[0] = Shape()
	graphics[1] = Dot()
	graphics[2] = Rectangle()
	graphics[3] = Circle()

	graphics.forEach { it?.accept(visitor) }
}